src/components/SearchBar.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from 'react';
+import * as _ from 'underscore';
+
+export interface SearchBarProps {
+    onTextChange: Function;
+    onStructuredModeChanged: Function;
+    onDetailedModeChanged: Function;
+    onCategoriesChange: Function;
+    categories: string[];
+}
```

Comment:
TASK: turn all `Function` usage here into some function prototypes that outline the parameters/return types.